owlkettle/playground.nim

```diff
@@ -20,83 +20,86 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 # SOFTWARE.
 
-import std/[options, times, macros, strformat, strutils, sequtils, typetraits]
+import std/[options, times, macros, strformat, sugar, strutils, sequtils, typetraits, tables]
 import ./dataentries
 import ./adw
 import ./widgetutils
 import ./guidsl
 import ./widgetdef
 import ./widgets
 
-macro getField*(someType: untyped, fieldName: static string): untyped =
-  nnkDotExpr.newTree(someType, ident(fieldName))
+type Range = concept r # Necessary as there is no range typeclass *for parameters*. So `field: ptr range` is not a valid parameter.
+  r is range
 
-# Default `toFormField` implementations
-proc toFormField[T: SomeNumber](state: auto, fieldName: static string, typ: typedesc[T]): Widget =
-  ## Provides a form field for all number times in SomeNumber
+# Forward Declarations so order of procs below doesn't matter
```

Comment:
Forward Declarations here mostly exist for the procs with  `ptr seq[T]` and `ptr auto`.
I want both of them to have access to *all* other `toFormField` procs, including each other.
This was the only way I could think of that enabled this.